core/silkworm/consensus/clique/clique_snapshot.cpp

```diff
@@ -0,0 +1,256 @@
+            recents_.pop_back();
+            // Update Tallies
+            clear_votes(header.beneficiary);
+        }
+        if (tallies_.count(header.beneficiary)) {
+            tallies_.erase(header.beneficiary);
+        }
+    }
+
+    update(header.number, hash);
+    // Success
+    return ValidationResult::kOk;
+}
+
+//! \brief Verify seal for header
+//! \param header: header to verify.
+ValidationResult CliqueSnapshot::verify_seal(const BlockHeader& header, const evmc::address& signer) {
+    // Check difficulty
+    if (is_authority(header.number, signer) && header.difficulty != kDiffInTurn && header.difficulty != kDiffNoTurn) {
+        return ValidationResult::kInvalidSeal;
+    }
+
+    return ValidationResult::kOk;
+}
+
+//! \brief Checks for authority
+//! \param block_number: Block to check.
+//! \param address: Address to check.
+//! \return if a signer at a given block height is in charge or not.
+bool CliqueSnapshot::is_authority(uint64_t block_number, evmc::address address) const noexcept {
+    if (signers_.empty()) {
+        return false;
+    }
+    return signers_[block_number % signers_.size()] == address;
+}
+
+//! \brief Getter method for signers_.
+//! \return Snapshot's signers.
+const std::vector<evmc::address>& CliqueSnapshot::get_signers() const noexcept { return signers_; }
+//! \brief Getter method for recents_.
+//! \return Snapshot's recents.
+const std::deque<evmc::address>& CliqueSnapshot::get_recents() const noexcept { return recents_; }
+
+//! \brief Getter method for block_number_.
+//! \return Snapshot's block number.
+const uint64_t& CliqueSnapshot::get_block_number() const noexcept { return block_number_; }
+
+//! \brief Getter method for hash_.
+//! \return Snapshot's hash.
+const evmc::bytes32& CliqueSnapshot::get_hash() const noexcept { return hash_; }
+
+//! \brief Convert the snapshot in Bytes.
+//! \return The resulting Bytes.
+Bytes CliqueSnapshot::to_bytes() const noexcept {
+    auto signers_bytes_count{signers_.size() * kAddressLength};
+    auto recents_bytes_count{recents_.size() * kAddressLength};
+    Bytes ret(signers_bytes_count + recents_bytes_count + 8 /* byte for signers count*/, '\0');
+
+    // We specify how many are signers
+    size_t offset{0};
+    endian::store_big_u64(&ret[offset], signers_.size());
+    offset += sizeof(uint64_t);
+    std::memcpy(&ret[offset], signers_.data(), signers_bytes_count);
+    offset += signers_bytes_count;
+
+    // We add the recents
+    for (const auto& recent : recents_) {
+        std::memcpy(&ret[offset], recent.bytes, kAddressLength);
+        offset += kAddressLength;
+    }
+
+    return ret;
+}
+
+//! \brief Decode snapshot from bytes format.
+//! \return Decoded snapshot.
+CliqueSnapshot CliqueSnapshot::from_bytes(ByteView& b, const uint64_t& block_number,
+                                          const evmc::bytes32& hash) noexcept {
+    // Consume signers count
+    auto signers_count{endian::load_big_u64(&b[0])};
```

Comment:
We must check the case when `signers_.empty()` and/or `recents_.empty()` otherwise we get errors (index-out-bounds) in encoding/decoding tests. Specifically test "Recents without signers snapshot encoding/decoding"